WIP: Added support for a C API.

eaplatanios opened the pull request:
I am planning to use Gym Retro from Swift (in combination with Swift for TensorFlow) and so I implemented a C API. It's almost complete (except for a couple functions that I can add soon), but I wanted to get feedback on whether this looks ok to merge.

Thanks,
Anthony

christopherhesse commented:
Doesn't Swift for Tensorflow support python calls directly?  Is this a performance optimization for that?

eaplatanios commented:
You can invoke Python yes, but that has the extra overhead of going through the Python interpreter for every call and the requirement that Python and all relevant dependencies is installed. This PR introduces a C API that can be used directly from Swift (i.e., no Python overhead or requirements) and that can also be potentially used to interface with other languages (e.g., Java/Scala/etc through JNI).

christopherhesse commented:
Out of curiosity, what is the overhead for swift + retro?

eaplatanios commented:
If you mean by Swift + Python Retro, I haven't implemented the Swift interface yet to test, but I have benchmarked another simple RL environment we built in C++ between Python and Swift (using C API bindings similar to this PR) and the performance increase was more than 5-fold. Python is interpreted vs Swift is compiled, which slows things down a lot, plus the communication overhead: Swift -> Python -> Native is much slower than Swift -> Native, which is what this PR would allow.

Note also that his enables retro to be used from other languages than Swift and Python.

eaplatanios commented:
@christopherhesse This is now complete. The C API now supports all the functionality of the python interface. I plan to add a Swift API in a separate repository soon.

christopherhesse commented:
How does a line like https://github.com/openai/retro/pull/152/files#diff-ade9c892f04c242a8aa4fefdd15f8116R99 compile in C?

christopherhesse commented:
@endrift any objections to a C API for retro?

eaplatanios commented:
@christopherhesse The implementation is still C++, but the API is C-compatible. This means that the calling convention for all functions defined in the new header is C-compatible and can be linked to when compiling C code that uses them.

eaplatanios commented:
@christopherhesse Actually, I need to make a few more edits. I'll try to push them asap.

eaplatanios commented:
@christopherhesse Actually never mind, I just pushed the fixes.

eaplatanios commented:
@christopherhesse I now have a very early (working) draft implementation of a Swift API for Retro that uses this C API. You can find it [here](https://github.com/eaplatanios/retro-swift).

eaplatanios commented:
I also fixed a couple of minor bugs in `emulator.cpp` and `coreinfo.cpp` (the latter causing the issue in #153).

christopherhesse commented on src/utils.cpp:
```diff
@@ -80,4 +81,12 @@ string drillUp(const vector<string>& targets, const string& fail, const string&
 	return {};
 }
+
+const char* newCString(std::string value) {
```
is this different from `strdup()`?

eaplatanios replied:
You're right. I changed it to using `strdup()`.

endrift commented on CMakeLists.txt:
```diff
@@ -285,6 +286,19 @@ if(BUILD_PYTHON)
+        CXX_VISIBILITY_PRESET default)
+    if(WIN32)
+        if(CMAKE_SYSTEM_PROCESSOR STREQUAL "x86_64" OR CMAKE_SYSTEM_PROCESSOR STREQUAL "AMD64")
+            add_definitions(-DMS_WIN64)
```
Why is this needed here? We already have add_definitions for if you're building the Python bindings. That should probably be pulled out and shared.

endrift commented on src/coreinfo.cpp:
```diff
@@ -108,7 +108,7 @@ bool loadCoreInfo(const string& jsonData) {
 			s_extensionToCore[*ext] = core.key();
 		}
-		s_coreToLib[core.key()] = core->at("lib");
+		s_coreToLib[core.key()] = core->at("lib").get<string>();
```
Separate PR for this please.

eaplatanios replied:
Done in #155 .

endrift commented on src/emulator.cpp:
```diff
@@ -360,7 +360,7 @@ bool Emulator::cbEnvironment(unsigned cmd, void* data) {
 	}
 	case RETRO_ENVIRONMENT_GET_SYSTEM_DIRECTORY:
-		*reinterpret_cast<const char**>(data) = corePath().c_str();
+		*reinterpret_cast<const char**>(data) = strdup(corePath().c_str());
```
This seems leaky. Also, should probably be in a separate PR.

eaplatanios replied:
I only added this because I used valgrind and saw that you use this after it's deallocated which means probably outside the scope of this function.

endrift replied:
While this is true that means it leaks memory as is and should probably be cached somewhere.

eaplatanios replied:
Pulled out in #156 so we can continue the discussion there. Not sure who is supposed to own the memory pointed to by `data` and free it eventually, but the way this is currently implemented does not give me any valgrind errors whereas skipping the `strdup` I get multiple errors because it's used after being deallocated.

endrift commented on src/retro-c.cpp:
```diff
@@ -0,0 +1,476 @@
+using namespace Retro;
+
+CEmulator* emulatorCreate(const char* romPath) {
+  Retro::Emulator* emulator = new Retro::Emulator();
```
Please match the existing style.

endrift replied:
Also if you're using namespace Retro you don't need the Retro:: here.

eaplatanios replied:
I only used `Retro::` to try and match the style used in your Python API implementation. I removed that now, but what else do you refer to when you ask to match the existing style? Is it about renaming `emulator` to `m_re`?

endrift commented on src/retro-c.cpp:
```diff
@@ -0,0 +1,476 @@
+
+CEmulator* emulatorCreate(const char* romPath) {
+  Retro::Emulator* emulator = new Retro::Emulator();
+  if (Emulator::isLoaded()) {
```
Leaky. Do this before making the object.

endrift commented on src/retro-c.cpp:
```diff
@@ -0,0 +1,476 @@
+  }
+  if (!emulator->loadRom(romPath)) {
+    fprintf(stderr, "Could not load ROM.");
+    return nullptr;
```
Leaky. Free the object.

endrift commented on src/retro-c.cpp:
```diff
@@ -0,0 +1,476 @@
+  }
+  // The following is necessary because otherwise, you get a segfault 
+  // when you try to get the screen for the first time.
+  emulator->run();
```
This is an inconsistency between the C and C++ APIs and should be addressed elsewhere.

eaplatanios replied:
I only added this here to follow the convention you use in the Python API. You also have it there.

eaplatanios replied:
Should I just remove it and not follow what is done in the Python API?

endrift commented on src/retro-c.cpp:
```diff
@@ -0,0 +1,476 @@
+  // The following is necessary because otherwise, you get a segfault 
+  // when you try to get the screen for the first time.
+  emulator->run();
+  return new CEmulator {emulator, 0};
```
Please match existing style wrt spacing.

eaplatanios replied:
What do you mean here? Could you please clarify by showing an example of how these lines should look in the existing style?

endrift commented on src/retro-c.cpp:
```diff
@@ -0,0 +1,476 @@
+}
+
+void emulatorDelete(CEmulator* emulator) {
+  delete ((Retro::Emulator*) emulator->emulator);
```
`static_cast<Emulator>`.

eaplatanios replied:
I don't usually write C++ code so I'm curious, why is a static cast preferable to the C-style cast? According to [this](https://en.cppreference.com/w/cpp/language/explicit_cast), in this case they should be equivalent.

endrift commented on src/retro-c.cpp:
```diff
@@ -0,0 +1,476 @@
+}
+
+void emulatorStep(CEmulator* emulator) {
+  ((Retro::Emulator*) emulator->emulator)->run();
```
Ditto.

endrift commented on src/retro-c.cpp:
```diff
@@ -0,0 +1,476 @@
+}
+
+CBytes* emulatorGetState(CEmulator* emulator) {
+  size_t numBytes = ((Retro::Emulator*) emulator->emulator)->serializeSize();
```
Ditto.

endrift commented on src/retro-c.cpp:
```diff
@@ -0,0 +1,476 @@
+CBytes* emulatorGetState(CEmulator* emulator) {
+  size_t numBytes = ((Retro::Emulator*) emulator->emulator)->serializeSize();
+  void* bytes = malloc(numBytes);
+  ((Retro::Emulator*) emulator->emulator)->serialize(bytes, numBytes);
```
Ditto.

endrift commented on src/retro-c.cpp:
```diff
@@ -0,0 +1,476 @@
+  size_t numBytes = ((Retro::Emulator*) emulator->emulator)->serializeSize();
+  void* bytes = malloc(numBytes);
+  ((Retro::Emulator*) emulator->emulator)->serialize(bytes, numBytes);
+  return new CBytes {reinterpret_cast<const uint8_t*>(bytes), numBytes};
```
Ditto.

endrift commented on src/retro-c.cpp:
```diff
@@ -0,0 +1,476 @@
+}
+
+bool emulatorSetState(CEmulator* emulator, CBytes* state) {
+  return ((Retro::Emulator*) emulator->emulator)->unserialize(state->bytes, state->numBytes);
```
Ditto.

endrift commented on src/retro-c.cpp:
```diff
@@ -0,0 +1,476 @@
+}
+
+CEmulatorScreen* emulatorGetScreen(CEmulator* emulator) {
+  auto* emu = ((Retro::Emulator*) emulator->emulator);
```
Ditto.

endrift commented on src/retro-c.cpp:
```diff
@@ -0,0 +1,476 @@
+    in = Image(Image::Format::RGBX888, emu->getImageData(), w, h, emu->getImagePitch());
+  }
+  in.copyTo(&out);
+  return new CEmulatorScreen {data, (size_t) w, (size_t) h, 3};
```
Ditto.

endrift commented on src/retro-c.cpp:
```diff
@@ -0,0 +1,476 @@
+}
+
+double emulatorGetScreenRate(CEmulator* emulator) {
+  return ((Retro::Emulator*) emulator->emulator)->getFrameRate();
```
Ditto.

endrift commented on src/retro-c.cpp:
```diff
@@ -0,0 +1,476 @@
+}
+
+CEmulatorAudio* emulatorGetAudio(CEmulator* emulator) {
+  size_t numSamples = ((Retro::Emulator*) emulator->emulator)->getAudioSamples() * 2;
```
Ditto.

endrift commented on src/retro-c.cpp:
```diff
@@ -0,0 +1,476 @@
+CEmulatorAudio* emulatorGetAudio(CEmulator* emulator) {
+  size_t numSamples = ((Retro::Emulator*) emulator->emulator)->getAudioSamples() * 2;
+  int16_t* samples = new int16_t[numSamples];
+  memcpy(samples, ((Retro::Emulator*) emulator->emulator)->getAudioData(), numSamples * 2);
```
Ditto.

endrift commented on src/retro-c.cpp:
```diff
@@ -0,0 +1,476 @@
+  size_t numSamples = ((Retro::Emulator*) emulator->emulator)->getAudioSamples() * 2;
+  int16_t* samples = new int16_t[numSamples];
+  memcpy(samples, ((Retro::Emulator*) emulator->emulator)->getAudioData(), numSamples * 2);
+  return new CEmulatorAudio {samples, numSamples};
```
Ditto.

endrift commented on src/retro-c.cpp:
```diff
@@ -0,0 +1,476 @@
+}
+
+double emulatorGetAudioRate(CEmulator* emulator) {
+  return ((Retro::Emulator*) emulator->emulator)->getAudioRate();
```
Ditto.

endrift commented on src/retro-c.cpp:
```diff
@@ -0,0 +1,476 @@
+CEmulatorResolution* emulatorGetResolution(CEmulator* emulator) {
+  auto w = (size_t) ((Retro::Emulator*) emulator->emulator)->getImageWidth();
+  auto h = (size_t) ((Retro::Emulator*) emulator->emulator)->getImageHeight();
+  return new CEmulatorResolution {w, h};
```
Leaky. Return a struct, not a pointer to struct, since this is so small.

endrift commented on src/retro-c.cpp:
```diff
@@ -0,0 +1,476 @@
+    return;
+  }
+  for (int key = 0; key < maskSize; key++) {
+    ((Retro::Emulator*) emulator->emulator)->setKey(player, key, mask[key]);
```
Ditto.

endrift commented on src/retro-c.cpp:
```diff
@@ -0,0 +1,476 @@
+}
+
+void emulatorAddCheat(CEmulator* emulator, const char* code) {
+  ((Retro::Emulator*) emulator->emulator)->setCheat(emulator->cheats, true, code);
```
Ditto.

endrift commented on src/retro-c.cpp:
```diff
@@ -0,0 +1,476 @@
+}
+
+void emulatorClearCheats(CEmulator* emulator) {
+  ((Retro::Emulator*) emulator->emulator)->clearCheats();
```
Ditto.

endrift commented on src/retro-c.cpp:
```diff
@@ -0,0 +1,476 @@
+}
+
+void emulatorConfigureData(CEmulator* emulator, CGameData* gameData) {
+  ((Retro::Emulator*) emulator->emulator)->configureData((Retro::GameData*) gameData->data);
```
Ditto.

endrift commented on src/retro-c.cpp:
```diff
@@ -0,0 +1,476 @@
+}
+
+CMemoryView* memoryViewCreate(void* addressSpace) {
+  return new CMemoryView { (Retro::AddressSpace*) addressSpace };
```
Ditto.

endrift commented on src/retro-c.cpp:
```diff
@@ -0,0 +1,476 @@
+}
+
+int64_t memoryViewExtract(CMemoryView* memoryView, size_t address, const char* type) {
+  return (*((Retro::AddressSpace*) memoryView->addressSpace))[Variable{ type, address }];
```
Ditto.

endrift commented on src/retro-c.cpp:
```diff
@@ -0,0 +1,476 @@
+}
+
+void memoryViewAssign(CMemoryView* memoryView, size_t address, const char* type, int64_t value) {
+  (*((Retro::AddressSpace*) memoryView->addressSpace))[Variable{ type, address }] = value;
```
Ditto.

endrift commented on src/retro-c.cpp:
```diff
@@ -0,0 +1,476 @@
+    blocks[i] = {iter.first, static_cast<const uint8_t*>(iter.second.offset(0)), iter.second.size()};
+    i++;
+  }
+  return new CMemoryBlocks {blocks, numBlocks};
```
Leaky. There's no associated delete method.

endrift commented on src/retro-c.cpp:
```diff
@@ -0,0 +1,476 @@
+}
+
+CSearch* searchCreate(const char** types, size_t numTypes) {
+  Retro::Search* search;
```
`search = nullptr` obviates the else block.

endrift commented on src/retro-c.cpp:
```diff
@@ -0,0 +1,476 @@
+
+CSearchResult* searchUniqueResult(CSearch* search) {
+  TypedSearchResult result = ((Retro::Search*) search->search)->uniqueResult();
+  return new CSearchResult {result.address, result.type.type};
```
Leaky. No associated delete.

endrift commented on src/retro-c.cpp:
```diff
@@ -0,0 +1,476 @@
+      result.second.size()};
+    i++;
+  }
+  return new CSearchTypedResults {cResults, results.size()};
```
Leaky. No associated delete.

endrift commented on src/retro-c.h:
```diff
@@ -0,0 +1,191 @@
+} CEmulatorAudio;
+
+typedef struct CEmulatorResolution {
+  size_t width;
```
`int`

endrift commented on src/retro-c.h:
```diff
@@ -0,0 +1,191 @@
+
+typedef struct CEmulatorResolution {
+  size_t width;
+  size_t height;
```
`int`

endrift commented on src/utils.cpp:
```diff
@@ -2,6 +2,7 @@
 
 #include <climits>
 #include <cstdlib>
+#include <cstring>
```
Revert changes to this file please.

endrift commented on src/retro-c.cpp:
```diff
@@ -0,0 +1,476 @@
+}
+
+bool gameDataSave(CGameData* gameData, const char* dataFilename, const char* scenarioFilename) {
+  ScriptContext::reset();
```
Why is this here?

endrift commented on src/retro-c.cpp:
```diff
@@ -0,0 +1,476 @@
+    }
+    i++;
+  }
+  return new CValidActions {actions, numActionsInner, numActionsOuter};
```
Leaky. No associated delete.

endrift commented on src/retro-c.cpp:
```diff
@@ -0,0 +1,476 @@
+    namesArray[i] = strdup(var.first.c_str());
+    i++;
+  }
+  return new CNames {namesArray, allValues.size()};
```
Leaky. No associated delete.

endrift commented on src/retro-c.cpp:
```diff
@@ -0,0 +1,476 @@
+  auto* typeCopy = malloc(size);
+  memcpy(typeCopy, var.type.type, size);
+  auto* type = reinterpret_cast<const char*>(typeCopy);
+  return new CVariable {name, var.address, type};
```
Leaky. No associated delete.

endrift commented on src/retro-c.cpp:
```diff
@@ -0,0 +1,476 @@
+    variables[i] = {strdup(var.first.c_str()), v.address, v.type.type};
+    i++;
+  }
+  return new CVariables {variables, numVariables};
```
Leaky. No associated delete.

endrift commented on src/retro-c.cpp:
```diff
@@ -0,0 +1,476 @@
+  size_t width = 0;
+  size_t height = 0;
+  ((Retro::Scenario*) gameData->scenario)->getCrop(&x, &y, &width, &height, player);
+  return new CCropInfo {x, y, width, height};
```
Leaky. No associated delete.

endrift commented on src/retro-c.cpp:
```diff
@@ -0,0 +1,476 @@
+  for (int i = 0; i < names.size(); i++) {
+    namesArray[i] = strdup(names[i].c_str());
+  }
+  return new CNames {namesArray, names.size()};
```
Leaky. No associated delete.

eaplatanios commented:
Sorry the missing delete functions are because I didn't end up using these functions in the Swift API and forgot to add the deletes. I am actually curious, where are "searches" used? I didn't see them being used anywhere in the Python API.

Also, where do I use `free` for something allocated with `new`?

eaplatanios commented:
@endrift I believe I addressed all of your comments, but please tell if there is anything else I can address now. I also pulled out the couple of changes you requested in separate PRs.

christopherhesse commented:
Thanks for fixing those issues @eaplatanios!  I have a couple of questions:

1)  Seems like the tests should build the C API by default
2)  Would be nice to have a ctypes or cffi python test that loads the `retro-c` library and uses it to load a rom and perform a single step, just to make sure this doesn't get broken by accident, similar to https://github.com/openai/retro/blob/master/tests/test_env.py#L13

eaplatanios commented:
@christopherhesse Sorry for not responding here for so long but I got caught up with multiple conferences and with my thesis proposal. I'll get back to this very soon.

christopherhesse commented:
No rush, the code isn't changing quickly so the PR isn't going to be invalidated anytime soon.

eaplatanios commented:
I just got back to this and updated the other PR. Regarding your points:

1. Where should I make that change? As in, in which file of the repo?
2. I'm not familiar with `ctypes` or `cffi` and I don't have to add such a test now and so it may be a bit of time until I add such a test.

christopherhesse commented:
1) probably https://github.com/openai/retro/blob/master/travis.py#L81
2) Getting a basic ctypes thing working isn't too hard, and the advantage is that we wouldn't accidentally break the C API as easily later.  I'm not sure if it should be a requirement to merge this PR, but I'm leaning towards yes.  Here's a simple ctypes example: https://pgi-jcns.fz-juelich.de/portal/pages/using-c-from-python.html

eaplatanios commented:
Yes, that totally makes sense. I will look into it as soon as I get a chance. By the way, why do you have `s_loadedEmulator` be a static variable, thus only allowing one emulator instance per process? Wouldn't it be better to make it local to `Emulator` instances? Or is there a `libretro` limitation?

christopherhesse commented:
I am told that this is a libretro limitation, gym-retro contains code to raise exceptions if you try to load multiple emulators at the same time.

eaplatanios commented:
@christopherhesse fyi, I also implemented a Swift API for ALE and I observe a ~5x speedup over using retro. Note that this is a larger difference than you'd observe with atari-py and python retro as it is not yet using the latest version of ALE which gives a further performance boost.

christopherhesse commented:
That is a known issue: https://github.com/openai/retro/issues/127

@endrift's suggestion from here https://github.com/openai/retro/issues/172#issuecomment-512558040, which seems to be the best fix, is to use ALE's stella instead of the main stella.

endrift commented on CMakeLists.txt:
```diff
@@ -295,6 +296,18 @@ if(BUILD_C)
     target_link_libraries(retro-c retro-base ${STATIC_LDFLAGS})
+    install(TARGETS retro-c)
+    file(WRITE ${CMAKE_CURRENT_BINARY_DIR}/retro-c.pc
+"
```
I'd strongly recommend using `configure_file` and a .in file instead of doing this inline.

endrift commented:
@crypt096 Why did you leave a review? To the best of my knowledge you've never been involved with this project and my "strong recommendation" was never address?
